packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go

```diff
@@ -497,34 +525,98 @@ func (dhtPeer *DHTPeer) handleNewDelegationConnection(conn net.Conn) {
 
 	//linfo().Msgf("received a new connection from %s", conn.RemoteAddr().String())
 
-	// read agent address
+	// read agent registration message
 	buf, err := utils.ReadBytesConn(conn)
 	if err != nil {
-		lerror(err).Msg("while receiving agent's Address")
+		lerror(err).Msg("while receiving agent's registration request")
 		nbrConns.Dec()
 		return
 	}
 
-	addr := string(buf)
+	msg := &dhtnode.AcnMessage{}
+	err = proto.Unmarshal(buf, msg)
+	if err != nil {
+		lerror(err).Msg("couldn't deserialize acn registration message")
+		// TOFIX(LR) setting Msgs to err.Error is potentially a security vulnerability
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_SERIALIZATION, Msgs: []string{err.Error()}}
+		response := &dhtnode.AcnMessage{Version: dhtnode.CurrentVersion, Payload: &dhtnode.AcnMessage_Status{Status: status}}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytesConn(conn, buf)
+		ignore(err)
+
+		nbrConns.Dec()
+		return
+	}
+
+	linfo().Msgf("Received registration request %s", msg)
+
+	// Get Register message
+	var register *dhtnode.Register
+	switch pl := msg.Payload.(type) {
+	case *dhtnode.AcnMessage_Register:
+		register = pl.Register
+	default:
+		err = errors.New("Unexpected payload")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_UNEXPECTED_PAYLOAD, Msgs: []string{err.Error()}}
+		response := &dhtnode.AcnMessage{Version: dhtnode.CurrentVersion, Payload: &dhtnode.AcnMessage_Status{Status: status}}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytesConn(conn, buf)
+		ignore(err)
+
+		nbrConns.Dec()
+		return
+	}
+	record := register.Record
+	addr := record.Address
+
 	linfo().Msgf("connection from %s established for Address %s",
 		conn.RemoteAddr().String(), addr)
 
+	// check if the PoR is valid
+	myPubKey, err := utils.FetchAIPublicKeyFromPubKey(dhtPeer.publicKey)
+	ignore(err)
+	status, err := dhtnode.IsValidProofOfRepresentation(record, addr, myPubKey)
```

Comment:
[STP2] peer will verify agent PoR before registering it the agent and saves it locally